deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -52,25 +55,178 @@ def upload(options, screenshots, max_n_threads = 16)
         localizations = version.get_app_store_version_localizations
       end
 
-      upload_screenshots(screenshots_per_language, localizations, options, max_n_threads)
+      upload_screenshots(updated_sets_per_language, localizations, options, max_n_threads)
     end
 
-    def delete_screenshots(localizations, screenshots_per_language, max_n_threads, tries: 5)
+    def get_updated_screenshots(localizations, screenshots_per_language)
+      sets_per_language = {} # per locale and device type; all provided screenshots
+      updated_sets_per_language = {} # per locale and device type; just the updated screenshots
+
+      # first, divide the new screenshots into sets
+      screenshots_per_language.each do |language, screenshots_for_language|
+        screenshots_per_device_type = {}
+
+        screenshots_for_language.each do |screenshot|
+          display_type = screenshot.device_type
+
+          if display_type.nil?
+            UI.error("Error... Screenshot size #{screenshot.screen_size} not valid for App Store Connect")
+            next
+          end
+
+          bytes = File.binread(screenshot.path)
+          checksum = Digest::MD5.hexdigest(bytes)
+
+          screenshots_per_device_type[display_type] ||= []
+          screenshots_per_device_type[display_type] << {
+              screenshot: screenshot,
+              checksum: checksum
+          }
+        end
+
+        sets_per_language[language] = screenshots_per_device_type
+      end
+
+      # then, compare the new screenshots with the existing ones
+      sets_per_language.each do |language, screenshots_per_device_type|
+        updated_screenshots_per_device_type = {}
+        localization = localizations.find do |l|
+          l.locale == language
+        end
+
+        unless localization
+          UI.error("Couldn't find localization on version for #{language}")
+          next
+        end
+
+        app_screenshot_sets_map = {}
+        app_screenshot_sets = localization.get_app_screenshot_sets
+        app_screenshot_sets.each do |app_screenshot_set|
+          app_screenshot_sets_map[app_screenshot_set.screenshot_display_type] = app_screenshot_set
+        end
+
+        screenshots_per_device_type.each do |device_type, screenshots_with_checksums|
```

Comment:
Maybe I'd consider renaming some of these variables 🤔 it's a bit difficult to keep track of what variables refer to screenshots that have been passed to the method and which variables refer to data on App Store Connect. For example, it's impossible to understand just by its name that `screenshots_per_device_type` refers to the screenshots that we (might) want to upload, you need to trace back its definition.

Reply:
I agree. What would you suggest? Since the original also referred to them as `screenshots` even if it was the `Deliver` instance or the `Spaceship` instance.

Reply:
At one point, I was using `existing_screenshot` for the data that is already in App Store Connect. Does that work better?

Reply:
I don't really have a suggestion 😁 I think `existing_screenshot`, `remote_screenshot` or `app_store_screenshot` could all work fine to denote a previously uploaded screenshot. The tricky part is finding a name for the new ones, but `new` doesn't work because it suggests they should be _added_ (while in fact they might end up being identical). Maybe `candidate_screenshot` for the local ones?

Reply:
I changed the names of some variables. 😃 All the existing stuff should now be referred to as `app_store_...` while we have `candidate_...` for screenshots supplied by the user and `changed_...` for screenshots and sets that contain added, deleted or updated screenshots.